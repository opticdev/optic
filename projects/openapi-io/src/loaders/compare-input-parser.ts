import path from "path";
import { inGit, loadSpecFromBranch } from "./file-on-branch";

enum SpecVersionFrom {
  file,
  git,
}

interface FromGit {
  from: SpecVersionFrom.git;
  branch: string;
  name: string;
}

interface FromFile {
  from: SpecVersionFrom.file;
  filePath: string;
}

export function parseSpecVersion(raw: string): FromFile | FromGit {
  if (raw.includes(":")) {
    const [rev, name] = raw.split(":");
    // if (!rev || !name) throw new Error("invalid git rev:name input " + raw);
    return {
      from: SpecVersionFrom.git,
      name: name.startsWith("/") ? name.substring(1) : name,
      branch: rev,
    };
  } else {
    return {
      from: SpecVersionFrom.file,
      filePath: raw,
    };
  }
}
