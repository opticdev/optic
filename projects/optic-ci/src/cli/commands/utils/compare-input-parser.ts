import { OpenAPIV3 } from '@useoptic/openapi-utilities';

export enum SpecVersionFrom {
  file,
  git,
  empty,
}

interface FromGit {
  from: SpecVersionFrom.git;
  branch: string;
  name: string;
}

interface FromFile {
  from: SpecVersionFrom.file;
  filePath: string;
}

interface FromEmpty {
  from: SpecVersionFrom.empty;
  value: OpenAPIV3.Document;
}

export type SpecFromInput = FromFile | FromGit | FromEmpty;

export function parseSpecVersion(
  raw: string | undefined,
  defaultSpec: OpenAPIV3.Document
): SpecFromInput {
  if (raw) {
    if (raw.includes(':')) {
      const [rev, name] = raw.split(':');
      // if (!rev || !name) throw new Error("invalid git rev:name input " + raw);
      return {
        from: SpecVersionFrom.git,
        name: name.startsWith('/') ? name.substring(1) : name,
        branch: rev,
      };
    } else {
      return {
        from: SpecVersionFrom.file,
        filePath: raw,
      };
    }
  } else {
    return {
      from: SpecVersionFrom.empty,
      value: defaultSpec,
    };
  }
}
