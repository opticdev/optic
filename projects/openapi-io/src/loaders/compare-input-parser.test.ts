import { parseSpecVersion } from "./compare-input-parser";

it("can parse a file ", async () => {
  expect(parseSpecVersion("/path/to/spec.json")).toMatchSnapshot();
});

it("can parse a git rev-file ", async () => {
  expect(parseSpecVersion("main:/path/to/spec.json")).toMatchSnapshot();
  expect(parseSpecVersion("feature/1:spec.json")).toMatchSnapshot();
  expect(
    parseSpecVersion("feature/1/define-spc:path/to/spec.json")
  ).toMatchSnapshot();
});
