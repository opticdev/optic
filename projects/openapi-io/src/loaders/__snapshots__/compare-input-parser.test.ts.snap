// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can parse a file  1`] = `
{
  "filePath": "/path/to/spec.json",
  "from": 0,
}
`;

exports[`can parse a git rev-file  1`] = `
{
  "branch": "main",
  "from": 1,
  "name": "path/to/spec.json",
}
`;

exports[`can parse a git rev-file  2`] = `
{
  "branch": "feature/1",
  "from": 1,
  "name": "spec.json",
}
`;

exports[`can parse a git rev-file  3`] = `
{
  "branch": "feature/1/define-spc",
  "from": 1,
  "name": "path/to/spec.json",
}
`;
